### PR TITLE
feat(launch): add job link to wandb footer

### PR DESCRIPTION
**Pull request by bcsherma**
Fixes
-----
- Fixes WB-14103

Description
-----------

This PR adds a link to a created job in the wandb footer, if a job was created.

### Summary

No summary available (Limit exceeded: required to process 99227 tokens, but only 50000 are allowed per call)

### Walkthrough
No walkthrough available (Limit exceeded: required to process 99227 tokens, but only 50000 are allowed per call)

Testing
-------
How was this PR tested?

Checklist
-------
- [ ] Include reference to internal ticket "Fixes WB-NNNNN" and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/main/CONTRIBUTING.md#conventional-commits)

copilot:poem

### CODE REVIEW

**Review by KyleGoyette — COMMENTED** [review]
Just a thought on the alias being used. Generally looks good to me. I will admit, I don't have all of the required context on the interface method implementations here

File: tests/pytest_tests/system_tests/test_core/test_footer.py
Comment (bcsherma) [reply tests/pytest_tests/system_tests/test_core/test_footer.py:133]: I'm seeing some really weird behavior actually. Right now we are limited to what is returned from ArtifactSaver.save https://github.com/wandb/wandb/blob/700ed2305c8cc57e46d1aa2d94a67072e50dd013/wandb/sdk/internal/sender.py#L1495

 Seems like sometimes this returns "latest" and sometimes it returns a version index. From what I can tell: if the artifact version already exists it returns the index alias (e.g v2), otherwise it returns latest.

File: wandb/sdk/interface/interface_shared.py
Comment (kptkin) [inline wandb/sdk/interface/interface_shared.py:550]: you don't need it since you are using publish. In general we are trying to get away from communicate and eventual remove it completely.

File: wandb/sdk/wandb_run.py
Comment (kptkin) [inline wandb/sdk/wandb_run.py:3438]: could you move into it's own utility method?

Comment (bcsherma) [reply wandb/sdk/wandb_run.py:3438]: Easy change to make, but I would actually prefer that we have the run link and the job link together in the output. Is that okay?

File: wandb/sdk/interface/interface_shared.py
Comment (bcsherma) [reply wandb/sdk/interface/interface_shared.py]: Removed :)

Comment (bcsherma) [reply wandb/sdk/interface/interface_shared.py]: Okay I remember why I added this now, func-linux-s_grpc-py39  and some others are failing with:
```
Traceback (most recent call last):
  File "/mnt/ramdisk/.tox/func-grpc-py39/lib/python3.9/site-packages/wandb/sdk/wandb_run.py", line 2133, in _atexit_cleanup
    self._on_finish()
  File "/mnt/ramdisk/.tox/func-grpc-py39/lib/python3.9/site-packages/wandb/sdk/wandb_run.py", line 2381, in _on_finish
    job_info_handle = self._backend.interface.deliver_request_job_info()
  File "/mnt/ramdisk/.tox/func-grpc-py39/lib/python3.9/site-packages/wandb/sdk/interface/interface.py", line 857, in deliver_request_job_info
    return self._deliver_request_job_info(job_link)
  File "/mnt/ramdisk/.tox/func-grpc-py39/lib/python3.9/site-packages/wandb/sdk/interface/interface_grpc.py", line 453, in _deliver_request_job_info
    job_info_response = self._stub.JobInfo(job_info)
  File "/mnt/ramdisk/.tox/func-grpc-py39/lib/python3.9/site-packages/grpc/_channel.py", line 1030, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/mnt/ramdisk/.tox/func-grpc-py39/lib/python3.9/site-packages/grpc/_channel.py", line 910, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNKNOWN
	details = "Exception calling application: 'InterfaceRelay' object has no attribute '_communicate_job_info'"
	debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"Exception calling application: \'InterfaceRelay\' object has no attribute \'_communicate_job_info\'", grpc_status:2, created_time:"2023-06-29T20:51:13.969197217+00:00"}"
>
INFO: exit= 255
```

File: wandb/sdk/wandb_run.py
Comment (kptkin) [reply wandb/sdk/wandb_run.py:3438]: just put the prints one after the other in the footer function?

File: wandb/sdk/interface/interface_shared.py
Comment (kptkin) [reply wandb/sdk/interface/interface_shared.py]: You need to fix `JobInfo` as well, you are calling communicate there.

File: wandb/sdk/wandb_run.py
Comment (bcsherma) [reply wandb/sdk/wandb_run.py:3438]: The thing is that this _footer_sync_info includes a line with output before and after the line with the run link, so I think I would need to break this function up even more or change the order that it is outputting the lines if I want to put the job link in its own helper while keeping it adjacent to the run link.

Comment (kptkin) [reply wandb/sdk/wandb_run.py:3438]: yeah, so keep it like this... you could just re-order things, but that's fine.

File: tests/pytest_tests/system_tests/test_core/test_footer.py
Comment (KyleGoyette) [reply tests/pytest_tests/system_tests/test_core/test_footer.py:133]: Mmm, that's not great, is the length ever > 1? Can we validate this in the core repo?

**Review by kptkin — APPROVED** [review]

File: wandb/sdk/internal/sender.py
Comment (kptkin) [inline wandb/sdk/internal/sender.py:301]: why it's not part of the job_builder attributes?

Comment (bcsherma) [reply wandb/sdk/internal/sender.py:301]: Good question, I will move them there

File: tests/pytest_tests/system_tests/test_core/test_footer.py
Comment (bcsherma) [reply tests/pytest_tests/system_tests/test_core/test_footer.py:133]: https://github.com/wandb/wandb/blob/3dc3fe5c4b01d18f248cc6f304bb0cf78c7e399b/wandb/sdk/internal/internal_api.py#L3168-L3178 the client does some stuff here to set its version key (which we use to print the link) to the version number if the version number is included in this other part of the response. But for some reason the version number isn't included in the response for v0 of the artifact, not sure why that is. I'm taking a look at the mutation but I don't have my bearings that well so it will take a bit

Comment (KyleGoyette) [reply tests/pytest_tests/system_tests/test_core/test_footer.py:133]: Then let's swap latest for v0, if multiple people run code that makes the same job collection, then latest may misdirect some users

File: wandb/sdk/internal/sender.py
Comment (KyleGoyette) [inline wandb/sdk/internal/sender.py:336]: Why is this change required by us here?

File: wandb/sdk/internal/sender.py
Comment (KyleGoyette) [inline wandb/sdk/internal/sender.py:383]: if we're not using record, is this necessary?

File: wandb/sdk/lib/printer.py
Comment (KyleGoyette) [inline wandb/sdk/lib/printer.py:200]: nit: this lightning is not like the other emojis?

File: wandb/sdk/internal/sender.py
Comment (KyleGoyette) [inline wandb/sdk/internal/sender.py:1520]: I'm a little concerned about this, are we sure the response always has these? Should we wrap this in a try, except to prevent users runs from crashing?

**Review by KyleGoyette — COMMENTED** [review]
Generally looks great! One question about a try except that I want to clear up before we merge

File: wandb/sdk/internal/sender.py
Comment (kptkin) [reply wandb/sdk/internal/sender.py:336]: agreed! you shouldn't change it!

File: wandb/sdk/internal/sender.py
Comment (kptkin) [reply wandb/sdk/internal/sender.py:1520]: also i assumed that when you move these attributes to the job_builder you would also move this logic there...

File: wandb/sdk/internal/sender.py
Comment (bcsherma) [reply wandb/sdk/internal/sender.py:336]: This is a merging issue

File: wandb/sdk/internal/sender.py
Comment (kptkin) [inline wandb/sdk/internal/sender.py:1073]: why??? this fix is here to avoid hangs! don't remove it!

Comment (bcsherma) [reply wandb/sdk/internal/sender.py:1073]: I've accidentally made a bunch of reverts here. I will audit all the changes in the PR right now.

File: wandb/sdk/interface/interface.py
Comment (kptkin) [inline wandb/sdk/interface/interface.py:856]:
```suggestion
        job_info = pb.JobInfoRequest()
```

File: wandb/sdk/interface/interface.py
Comment (kptkin) [inline wandb/sdk/interface/interface.py:860]:
```suggestion
    def _deliver_request_job_info(self, job_info: pb.JobInfoRequest) -> MailboxHandle:
```

File: wandb/sdk/internal/sender.py
Comment (kptkin) [inline wandb/sdk/internal/sender.py:734]: should the default values or the result of the getter be also part of the job_builder?

File: wandb/sdk/interface/interface.py
Comment (kptkin) [inline wandb/sdk/interface/interface.py:857]:
```suggestion
        return self._deliver_request_job_info(job_info)
```

**Comment by kptkin** [comment]
@bcsherma could you also try it in offline mode (ideally actual offline, i.e. air-gapped)?

**Review by kptkin — CHANGES_REQUESTED** [review]
i think we should verify this works in offline mode, preferably air-gapped...

**Comment by bcsherma** [comment]
> i think we should verify this works in offline mode, preferably air-gapped...

@kptkin I went ahead and tried with in offline mode with my machine offline. There were no errors. After coming back on line I was able to `wandb sync` the run without any errors and the job is synced correctly. Should I add a test for this?

**Review by kptkin — APPROVED** [review]

**Comment by kptkin** [comment]
i'll let Kyle finish the review!

File: wandb/sdk/internal/job_builder.py
Comment (KyleGoyette) [inline wandb/sdk/internal/job_builder.py:115]: nit: this reads weird to me, I'd do: if type==job; try/except, elif type==code...

File: wandb/sdk/lib/printer.py
Comment (KyleGoyette) [inline wandb/sdk/lib/printer.py:200]: nit: lightning weird again?

**Review by KyleGoyette — APPROVED** [review]
Approved with some nits

File: wandb/sdk/lib/printer.py
Comment (bcsherma) [reply wandb/sdk/lib/printer.py:200]: I think this is just something with how that character is displayed in github's code view. If you copy and paste that character elsewhere it looks like ⚡

File: wandb/sdk/internal/job_builder.py
Comment (bcsherma) [reply wandb/sdk/internal/job_builder.py]: done

**Pull request merged**
